crates/llm-local/src/llama.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{token_output_stream, utils::load_safetensors, CachedInferencingModel};
+use anyhow::{anyhow, Result};
+use candle::{utils, Device, Tensor};
+use candle_nn::VarBuilder;
+use candle_transformers::{
+    generation::{LogitsProcessor, Sampling},
+    models::llama::{self, Cache, Config, Llama, LlamaConfig},
+};
+use rand::{RngCore, SeedableRng};
+use spin_core::async_trait;
+use spin_world::v2::llm::{self as wasi_llm, InferencingUsage};
+use std::{fs, path::PathBuf, sync::Arc};
+use tokenizers::Tokenizer;
+
+const TOKENIZER_FILENAME: &str = "tokenizer.json";
+const CONFIG_FILENAME: &str = "config.json";
+const EOS_TOKEN: &str = "</s>";
+const MODEL_SAFETENSORS_INDEX: &str = "model.safetensors.index.json";
+
+pub fn auto_device() -> Result<Device> {
+    if utils::cuda_is_available() {
+        Ok(Device::new_cuda(0)?)
+    } else if utils::metal_is_available() {
+        Ok(Device::new_metal(0)?)
+    } else {
+        Ok(Device::Cpu)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct LlamaModels {
+    model: Arc<Llama>,
+    config: Config,
+    cache: Box<Cache>,
```

Comment:
Do we want to clone the cache every time we clone `LlamaModels`?

Reply:
removed the usage of `Box` on cache